File: exercise/scala/day13/src/main/scala/santamarket/model/ReceiptItem.scala
package santamarket.model

import java.lang.Math.*

case class ReceiptItem(product: Product, quantity: Double, price: Double, totalPrice: Double):
  override def equals(obj: Any): Boolean = obj match
    case that: ReceiptItem =>
      abs(this.price - that.price) < 0.001 &&
        abs(this.totalPrice - that.totalPrice) < 0.001 &&
        abs(this.quantity - that.quantity) < 0.001 &&
        this.product == that.product
    case _ => false

  override def hashCode(): Int =
    (product, price, totalPrice, quantity).hashCode

File: exercise/C#/day13/SantaMarket/Model/ReceiptItem.cs
namespace SantaMarket.Model
{
    public class ReceiptItem
    {
        public ReceiptItem(Product product, double quantity, double price, double totalPrice)
        {
            Product = product;
            Quantity = quantity;
            Price = price;
            TotalPrice = totalPrice;
        }

        public double Price { get; }

        public Product Product { get; }

        public double Quantity { get; }

        public double TotalPrice { get; }

        public override bool Equals(object? obj)
        {
            if (ReferenceEquals(this, obj)) return true;
            if (obj is null || GetType() != obj.GetType()) return false;

            var that = (ReceiptItem) obj;
            const double tolerance = 0.001;

            return Math.Abs(that.Price - Price) < tolerance &&
                   Math.Abs(that.TotalPrice - TotalPrice) < tolerance &&
                   Math.Abs(that.Quantity - Quantity) < tolerance &&
                   Equals(Product, that.Product);
        }

        public override int GetHashCode() => HashCode.Combine(Product, Price, TotalPrice, Quantity);

        public override string ToString() =>
            $"ReceiptItem{{product={Product}, price={Price}, totalPrice={TotalPrice}, quantity={Quantity}}}";
    }
}